File: rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        int count = 0;
        ListNode *temp = head;
        while(temp!=nullptr){
            count++;
            temp= temp->next;
        }
        if(count <= 1 || k ==0){
            return head;
        }
        temp = head;
        k = k % count;
        if (k==0){
            return head;
        }
        while(temp->next!=nullptr){
            temp = temp->next;
        }
        temp->next = head;
        ListNode *temp1 = head, *temp2 = head->next;
        for(int i=0; i< count-k-1; i++){
            temp2 = temp2->next;
            temp1 = temp1->next;
        }head = temp2;
        temp1->next = nullptr;
        return head;
    }
};

File: rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) return head;
        ListNode temp = head;
        int count = 1;
        while(temp.next != null){
            temp = temp.next;
            count++;
        }temp.next = head;
        k = k%count;
        ListNode prev = head;
        for(int i = 1; i < count - k; i++){
            prev = prev.next;
        }ListNode nHead = prev.next;
        prev.next = null;
        head = nHead;
        return head;
    }
}